src/main/java/com/treasuredata/underwrap/UnderwrapMetrics.java

```diff
@@ -0,0 +1,80 @@
+package com.treasuredata.underwrap;
+
+import org.xnio.XnioWorker;
+
+import java.lang.reflect.Field;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+
+public class UnderwrapMetrics
+{
+    private final XnioWorker worker;
+
+    public UnderwrapMetrics(XnioWorker worker)
+    {
+        this.worker = worker;
+    }
+
+    // Latest Undertow (1.4.1x) depends on Xnio 3.3.x, and this version doesn't have methods to
+    // get metrics (for example, `XnioWorker#getMXBean()`).
+
+    // Methods below forcibly get metric values from XnioWorker internal, instead of calling
+    // protected/unexisting methods of XnioWorker.
+    // https://github.com/xnio/xnio/blob/3.x/api/src/main/java/org/xnio/XnioWorker.java#L889-L923
+
+    public int getCoreWorkerPoolSize()
+    {
+        // return coreSize;
+        try {
+            Field field = XnioWorker.class.getDeclaredField("coreSize");
+            field.setAccessible(true);
+            return field.getInt(worker);
+        }
+        catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException("BUG: incompatible accesse", e);
```

Comment:
Hmm... I'm not sure this method should throw an RuntimeException.
Maybe services should keep working healthy even if it fails to collect metrics. Of course, it depends, though.

Reply:
In this case, these exceptions (about reflection) will be caused by name of private fields or its types. These things are fixed with version of dependencies (Undertow and Xnio), and NOT be determined at runtime actually.
IMO, these things can be checked for such failures by CI on this repository. RuntimeException will be never thrown in actual applications.